modules/slugs/includes/class-qtranslate-slug.php

```diff
@@ -471,6 +464,19 @@ function filter_request( $query ) {
                     $function             = 'get_term_link';
                 }
             endforeach;
+
+             // -> post
+            if ( ! $function && ( isset($query['name']) || isset($query['p'] ) ) ) {
+                $post = isset($query['p']) ? get_post($query['p']) : $this->get_page_by_path($query['name'], OBJECT, 'post');
+                if (!$post) {
+                    return $query;
+                }
+                $query['name'] = $post->post_name;
+                $id = $post->ID;
+                $cache_array = array($post);
+                update_post_caches($cache_array);
+                $function = 'get_permalink';
+            }
```

Comment:
Why is that if-block nested inside the last else? The checks are not clear, it's getting too confusing. Is that for custom post type or more? Nested "if" should be really justified. It should be rewritten with very clear if sections. I preferred the former version with the "post" in the previous place.

Reply:
Problem is that `name` key is passed in `$query` arg for pages, posts and custom post types at least.
Need to perform different checks on other keys or `matched_query` to identify which is the correct callback.
As `post` seems to involve the more generic test, it is left as last fallback if no other function has been identified from the previous tests.
Moving it in the previous position would always give priority on the custom post types, which will result in 404.
If we revert everything to the original code, `$query` is completely discarded braking other possible custom query vars (only `matched_query `was used).